### DISCUSSION TITLE
New Grafana Chart - Monthly Stats

### DISCUSSION THREAD

**Discussion by pratik5705** [teslamate-org/teslamate#664] [Ideas]
**Is your feature request related to a problem? Please describe.**
I was trying to figure out my estimated miles driven per month and how that has changed with COVID.

Would it be possible to add some monthly stats to the grafana chart, broken down by calendar month:
Miles Driven
Charges
Charging Cost
Fuel Efficiency

I am sure there are others. 

Love this project, amazing work. Keep up the great work.

**Comment by cwanja** [comment]
Tried using the Trip dashboard and setting the time to 30 (or 31) days? Should include every thing you are after.

**Comment by pratik5705** [comment]
Thanks for the tip! Yes, it is. However, I was looking for something that
is comparative - compare April to March, etc. Would that have too many
variations?

Pratik

**Comment by Dulanic** [comment]
So I began working on it... but I dont like the layout yet..... Feel free to make some layout recommendations ... because this isn't doing it for me so far.

![image](https://user-images.githubusercontent.com/2948206/80738339-7c3b8300-8ada-11ea-992d-7e0fa5f7660b.png)

**Comment by cwanja** [comment]
Not the _best_ solution, but could you just open two tabs or windows and compare it? Are you really hitting this dashboard daily or weekly, or just once a month? If the latter, opening two tabs or windows seems like the best solution (in my opinion). Not saying you could not get some dashboard to eventually do it. Just wonder if it is worth the effort.

**Comment by pratik5705** [comment]
So I started with the Trips dashboard and removed some of the panels and duplicated the first panel (total distance). Using time shift property of the panel, I was able to get last 30 days and last 60-30days to show up.

First screenshot showing the 2 different values (time range set to last 30 days):
![image](https://user-images.githubusercontent.com/4359069/80747913-2377e480-8af2-11ea-927a-88831053badd.png)

Second screenshot confirming the right value shows up (time range set to now-60d & now-30d). However, the bottom panel also shows the same calculation.
![image](https://user-images.githubusercontent.com/4359069/80747717-dbf15880-8af1-11ea-99ce-27f801689fa6.png)

Screenshot showing the change to the relative time/time shift settings.
![image](https://user-images.githubusercontent.com/4359069/80748221-95502e00-8af2-11ea-999f-7c52fa96b9f2.png)


I am going to duplicate all the panels on the top to the bottom and potentially a third time to show
last 30 days
last 60-30 days
last 90-60 days

If I save this dashboard as a copy, will it get overwritten on the next update (i.e., when I update all the containers for the next version of TeslaMate?)
Thanks.

**Comment by pratik5705** [comment]
Ok, here is the copy with 2 sections showing last 30 days and last 60-30 days. Figured out the have to leave relative time blank and time shift 30d in the second set. Set the time filter to last 30 days and this works correctly.

![image](https://user-images.githubusercontent.com/4359069/80748892-a3eb1500-8af3-11ea-8888-adb65f349e59.png)

**Comment by pratik5705** [comment]
Ok, it is complete.

![image](https://user-images.githubusercontent.com/4359069/80751205-64bec300-8af7-11ea-8357-e408c9c805d7.png)

I am posting the JSON file (zipped) if anyone is interested in making this. 
[TripCopy.zip](https://github.com/adriankumpf/teslamate/files/4560794/TripCopy.zip)

Thanks community for showing me the way to do it!

**Comment by DrMichael** [comment]
Hmm, I thought, you meant more something like this?
![Trip-Monatsstatistik](https://user-images.githubusercontent.com/8087682/80806027-d5c6af00-8bb9-11ea-8843-988a74c809cd.png)

Could we learn from Teslalogger? :-)

**Comment by pratik5705** [comment]
That is what I was initially thinking but cwanja's tip to use the trip
dashboard, led me down a different path. Can you share the json for the
dashboard? I would like to see the query.

Pratik

**Comment by DrMichael** [comment]
Sure: https://github.com/bassmaster187/TeslaLogger/blob/master/TeslaLogger/Grafana/Trip%20Monatsstatistik.json

The query is:
SELECT DATE_FORMAT(StartDate, '%Y / %m') as date, \nsum(DurationMinutes)/60 as sum_duration_h, \nsum(km_diff) as sum_km,\nsum(consumption_kWh) as sum_consumption_kwh, \nsum(consumption_kWh) / sum(km_diff) * 100 as avg_consumption_kwh,\navg(outside_temp_avg) as avg_outside_temp,\ncount(*) as cnt\nfrom trip WHERE \n  $__timeFilter(StartDate) and km_diff < 1000 \n  group by DATE_FORMAT(StartDate, '%Y-%m')\n

**Comment by pratik5705** [comment]
I am working on converting this to work with Teslamate. Have the base
working, need to do the unit conversion. The consumption values are not in
the drives table, will look at the drive chart to calculate these values.

Will submit the completed dashboard once done.

Thanks,
Pratik

**Comment by pratik5705** [comment]
Here is the ported over dashboard from the above linked TeslaLogger dashboard. I didn't do the kWh/100 mi calculation. I can work on it, if you are interested in including this to TeslaMate. Also, some formatting might need to be checked on for km units.
Thanks again for sharing a path forward.
[Trip Stats Copy-1588370071040.zip](https://github.com/adriankumpf/teslamate/files/4566423/Trip.Stats.Copy-1588370071040.zip)

**Comment by DrMichael** [comment]
I also worked a bit on this... Please comment. And please check C/F and km/mi...

(I am not happy yet with the units of the "Distance driven".)

(Newer version available)

**Comment by DrMichael** [comment]
I would like to put links behind the "Distance driven" and "Energy charged" fields to jump to "Drives" and "Charges" respectively for the given month. Can anybody help me to get the time interval in the link?

It should be something like `from=${__cell_0}&to=${__cell_0}+1m`, (cell 0 holding the milliseconds since epoch), but the "+1m" does not work... And, apparently the time has to be adjusted to local time.

Any ideas?

**Comment by DrMichael** [comment]
Ok, I got further. Any comments? Is it useful anyway? :-)

![2020-06-04 11_34_06-Microsoft Edge](https://user-images.githubusercontent.com/8087682/83740663-73991800-a657-11ea-9358-3b3c84da3086.png)

New version with car selector:
(Newer version available)

**Comment by Dulanic** [comment]
Looks good, just needs a few updates. 

Very few languages use Ø . Not sure if I would hard code that. 

Also, my isn't formatted.... Oh I see, you added mi, but you set it for sum_distance_km, need to update to sum_distance_mi.

I wonder if we should consider averaging the "Efficiency", quoted since that's still up for discussion as to the right wording for that. At least the avg of it.

**Comment by DrMichael** [comment]
> Very few languages use Ø . Not sure if I would hard code that.

Took that out...

> Also, my isn't formatted.... Oh I see, you added mi, but you set it for sum_distance_km, need to update to sum_distance_mi.

Right.

> I wonder if we should consider averaging the "Efficiency", quoted since that's still up for discussion as to the right wording for that. At least the avg of it.

I am not sure about that either. But, it is about **energy** added... I used _charge_energy_used_, not sure, whether that is the right one. I had also to take out charges with charge_energy_used==0. (That I get, when the battery is full. Then once in a while it will add 0.1-0.2 kWh (charge_energy_added), but charge_energy_used is zero.)

**Comment by DrMichael** [comment]
Newer version available

![2020-06-04 18_25_35-FritzBoxTraffic](https://user-images.githubusercontent.com/8087682/83784886-e377c480-a690-11ea-821b-f0e43e980e4d.png)


> Coming along! I just imported this dashboard. I had 1000 free super charging miles I started to use up. These charges aren't showing up for me.
> 
> I had also opened up this [request](https://github.com/adriankumpf/teslamate/issues/724) that may tie along this one perhaps?

**Comment by natrlhy** [comment]
Coming along! I just imported this dashboard. I had 1000 free super charging miles I started to use up. These charges aren't showing up for me.

I had also opened up this [request](https://github.com/adriankumpf/teslamate/issues/724) that may tie along this one perhaps?

**Comment by DrMichael** [comment]
> I had 1000 free super charging miles I started to use up. These charges aren't showing up for me.

What do you mean with that? Are not showing up where?!?!
Can you check it with this version and explain?

Newer version available

**Comment by natrlhy** [comment]
Here are my charges:
![image](https://user-images.githubusercontent.com/19696424/83888948-620b4b00-a6ff-11ea-825f-43b88a7f1115.png)

I was using your Monthly _ yearly statistics for drives and charges-1591277192096.zip version and here is what I see:

![image](https://user-images.githubusercontent.com/19696424/83890377-8ddb0080-a700-11ea-9105-56e6cf0ff10b.png)

My screenshot above seems to show that the super charges aren't showing up in the monthly/yearly stats dashboard and version mentioned above.

I also tried your new version and looks the same as the screenshot provided

**Comment by DrMichael** [comment]
@natrlhy Could we jump on discord? https://discord.com/channels/631168014540341268/631168014540341270

**Comment by DrMichael** [comment]
@natrlhy Can you look at the database (`docker-compose exec database psql -U teslamate -d teslamate`)? What do you get with

`select start_date, end_date, charge_energy_added from charging_processes where start_date BETWEEN '2020-05-01T00:00:00.000Z' AND '2020-06-01T00:00:00.000Z';`

**Comment by DrMichael** [comment]
Anyway, new version...

[Statistics-1591372901660.zip](https://github.com/adriankumpf/teslamate/files/4737239/Statistics-1591372901660.zip)

**Comment by Dulanic** [comment]
For charge_energy, for vehicle usage you'd want charge_energy_added as that is the amount of power added to the batteries. If you want to see a "guess" as to how much energy was taken from the grid, that is charge_energy_used. I'd recommend just using charge_energy_added .

So this was my idea /w efficiency....

https://pastebin.com/Wx0ENLdz

**Comment by DrMichael** [comment]
Well, I'd like to get energy taken from the grid... That is, what I have consumed.

Efficiency is interesting, but a bit academic IMHO. And we have an Efficiency dashboard already.

But, do you have an idea, how to deal with the timezones? It appears, that they are stored in UTC. But, when displayed in Grafana, they are in local time. And I would like to have the month started at the first day 00:00 local time, not UTC time...

**Comment by DrMichael** [comment]
> So this was my idea /w efficiency....

@Dulanic I never looked into that efficiency thing, what exactly does it tell me? And what do the colors tell? Why did you color the Distance driven?

Do you perhaps have a solution for the UTC/local timezone issue? All dates showed in Grafana are essentially in UTC, meaning for UTC-7 the e.g. month starts on the evening of the last day of the previous month... Not good.

**Comment by DrMichael** [comment]
Well, this is as close as I can get with the timezone issue: the dashboard displays in UTC. But for that it should provide reasonable data.

[Statistics-1591537570944.zip](https://github.com/adriankumpf/teslamate/files/4742168/Statistics-1591537570944.zip)

**Comment by Dulanic** [comment]
Databases and timezones get to be fun :) So grafana will convert the timezone if it can, to make this possible, I recommend using epoch. 

As to why I colored the distance... mistake.

EX: for date field so it can convert can be done like so: extract(epoch from date_trunc('$period', start_date)) * 1000

Other options are to convert the timezone see below for examples:

SELECT date, timezone('UTC',date) , date AT TIME ZONE current_setting('TIMEZONE'), timezone('UTC',date) AT TIME ZONE current_setting('TIMEZONE')
from charges

**Comment by DrMichael** [comment]
> As to why I colored the distance... mistake.

:-) I like the color...

> EX: for date field so it can convert can be done like so: extract(epoch from date_trunc('$period', start_date)) * 1000

The problem is to adjust epoch for local timezone also in retrospect. So, I am in +2 currently and +1 in winter. I don't see a feasible solution...

> SELECT date, timezone('UTC',date) , date AT TIME ZONE current_setting('TIMEZONE'), timezone('UTC',date) AT TIME ZONE current_setting('TIMEZONE') from charges

Gives all the same. Well, the containers (postgres and grafana) are running with UTC and the postgres timezone is also UTC. Grafana concerts it then to local time, when that is set in the settings.

**Comment by Dulanic** [comment]
Ah yeah no I run my container in my timezones 😁

**Comment by Dulanic** [comment]
As a follow up, we could recommend people set their timezone /w the containers if we wanted that functionality. It's a easy flag....  EX: 

`-e TZ=America/Chicago`

**Comment by DrMichael** [comment]
https://stackoverflow.com/questions/44394295/how-to-change-timezone-of-postgresql-9-5-on-docker talks about setting TZ and PGTZ in the docker-compose.yml.

That works, but does not resolve adjusting a date to local time midnight because of DST.

@adriankumpf Any specific reason, why all containers except teslamate run in UTC?

**Comment by DrMichael** [comment]
Ok, I am running the database container and postgres now in local timezone.

Now, the following should do the job:
`date_trunc('month', TIMESTAMP WITH TIME ZONE start_date, current_setting('TIMEZONE')) as date`
(It does with '2001-02-16 20:38:40-05'...)

But I get `pq: syntax error at or near "start_date"`. What is wrong here?!?

**Comment by Dulanic** [comment]
Just wrong position/order.

```date_trunc('month', start_date::TIMESTAMP WITH TIME ZONE , current_setting('TIMEZONE')) as date```

or 

`date_trunc('month', start_date AT TIME ZONE current_setting('TIMEZONE')) as date`

**Comment by DrMichael** [comment]
Ah, ok. But now, when I put that into the SQL, I get
`pq: column "data.start_date" must appear in the GROUP BY clause or be used in an aggregate function`

Putting start_date in the GROUP BY gives not the desired result, aggregate start_date with e.g. min gives `pq: aggregate functions are not allowed in GROUP BY`.

**Comment by Dulanic** [comment]
Any time you use a aggregate, you need to group the non aggregate fields. If you put `date_trunc('month', start_date AT TIME ZONE current_setting('TIMEZONE'))` in the select, you need to put `date_trunc('month', start_date AT TIME ZONE current_setting('TIMEZONE'))` in the group by.

**Comment by DrMichael** [comment]
Hmmm, don't understand that... I assume, that I would need to put start_date in the GROUP BY, but I don't want that, because then I get a line for each day.

Could you perhaps have a look at the SQL query?

**Comment by DrMichael** [comment]
Ah, I was looking at the wrong place: all (other) start_date occurrences must be aggregated...
[Statistics-1591800199304.zip](https://github.com/adriankumpf/teslamate/files/4759279/Statistics-1591800199304.zip)

**Comment by Dulanic** [comment]
yeah worst case, if you don;'t want to, just aggregate the start_date with
min or max even if it's a individual line. Or purely group by the
date_part(month version.

**Comment by DrMichael** [comment]
Got better with taking the date_trunc into a WITH...
[Statistics-1591803816867.zip](https://github.com/adriankumpf/teslamate/files/4759677/Statistics-1591803816867.zip)

Timezone appears to switch to UTC sometimes though...

**Comment by Dulanic** [comment]
Looks right to me, not sure why it would switch to UTC for you. Especially if you set the timezone field to show current_setting('TIMEZONE') as Timezone, which should never change as it's stored in the DB settings and doesn't change.

**Comment by DrMichael** [comment]
So strange, I have not the faintest, where that "Etc/UTC" is still coming from. Simplified in Grafana it looks like:
![2020-06-11 11_11_39-Window](https://user-images.githubusercontent.com/8087682/84367554-b0738a80-abd4-11ea-90b2-84da5d243b84.png)
In psql it looks like:
![2020-06-11 11_12_04-Window](https://user-images.githubusercontent.com/8087682/84367577-b9fcf280-abd4-11ea-897a-2dd8a91bb7d9.png)
And `date` in the container gives me `Thu 11 Jun 2020 11:12:56 AM CEST`

**Comment by DrMichael** [comment]
Ok, it turned out, that although PGTZ is set to 'Europe/Berlin', /var/lib/postgresql/data/postgresql.conf contains 'Etc/UTC'. Changing that to 'Europe/Berlin' appears to fix the situation.

@adriankumpf Could we take care for that in the distributed container? Or much simpler approach: just store the local timezone in the settings...

@natrlhy Do you want to test it in UTC-7?

**Comment by Dulanic** [comment]
Oh that's easy, DST.

**Comment by DrMichael** [comment]
Hmmm, what do you mean?

**Comment by Dulanic** [comment]
Daylight savings start and end.

**Comment by DrMichael** [comment]
Yes, sure. But my question was more: why do I get Etc/UTC in Grafana and Europe/Berlin in psql: Because PGTZ sets the timezone for the **clients**.

**Comment by Dulanic** [comment]
Grafana must be doing it, just don't know why.

**Comment by DrMichael** [comment]
Ah, no.
The issue is (was), that PGTZ sets the timezone for psql. But it is not set, when Grafana issues the query. So, after setting the timezone to Europe/Berlin in postgresql.config, also Grafana shows and uses that timezone. (I hope, that does not break anything else... :-)

**Comment by DrMichael** [comment]
Ok, I share my latest version. Remember: you must change the timezone in the database container in `/var/lib/postgresql/data/postgresql.conf` to your local timezone!

(If you do not do it, it works with UTC, which leads to misleading entries in "Period"; at least for all UTC-N)

@adriankumpf Could we take care for that in the distributed container? Or simpler approach: just store the local timezone in the settings... Or set the timezone of the database to the local one?

[Statistics-1591949142513.zip](https://github.com/adriankumpf/teslamate/files/4769549/Statistics-1591949142513.zip)

**Comment by DrMichael** [comment]
Committed my latest version. Still the timezone issue exists...
